hazelcast/src/main/java/com/hazelcast/query/Predicates.java

```diff
@@ -465,6 +466,11 @@ public AndPredicate(Predicate... predicates) {
 
         @Override
         public Set<QueryableEntry> filter(QueryContext queryContext) {
+            Set<QueryableEntry> optoResult = tryOptimizedFiltering(queryContext);
+            if (optoResult != null) {
+                return optoResult;
+            }
+
             Set<QueryableEntry> smallestIndexedResult = null;
```

Comment:
This rewrite is done during execution. So a user has no idea that his query was rewritten. And even if it was.. I don't see a problem with that.